crates/supervisor/storage/src/providers/log_provider.rs

```diff
@@ -0,0 +1,385 @@
+//! Reth's MDBX-backed abstraction of [`LogProvider`] for superchain state.
+//!
+//! This module provides the [`LogProvider`] struct, which uses the
+//! [`reth-db`] abstraction of reth to store execution logs
+//! and block metadata required by the Optimism supervisor.
+//!
+//! It supports:
+//! - Writing full blocks of logs with metadata
+//! - Retrieving block metadata by number
+//! - Finding a block from a specific log (with hash/index match)
+//! - Fetching logs per block using dup-sorted key layout
+//!
+//! Logs are stored in [`LogEntries`] under dup-sorted tables, with log index
+//! used as the subkey. Block metadata is stored in [`BlockRefs`].
+
+use crate::{
+    error::StorageError,
+    models::{BlockRefs, LogEntries},
+};
+use kona_protocol::BlockInfo;
+use kona_supervisor_types::Log;
+use reth_db_api::{
+    cursor::{DbCursorRO, DbDupCursorRO, DbDupCursorRW},
+    transaction::{DbTx, DbTxMut},
+};
+use std::fmt::Debug;
+use tracing::{debug, error, warn};
+
+/// A log storage that wraps a transactional reference to the MDBX backend.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct LogProvider<'tx, TX> {
+    tx: &'tx TX,
+}
+
+/// Internal constructor and setup methods for [`LogProvider`].
+impl<'tx, TX> LogProvider<'tx, TX> {
+    #[allow(dead_code)]
+    pub(crate) const fn new(tx: &'tx TX) -> Self {
+        Self { tx }
+    }
+}
+
+#[allow(dead_code)]
+impl<TX> LogProvider<'_, TX>
+where
+    TX: DbTxMut + DbTx,
+{
+    fn store_block_logs(&self, block: &BlockInfo, logs: Vec<Log>) -> Result<(), StorageError> {
+        debug!(target: "supervisor_storage", block_number = block.number, "Storing logs");
+
+        if let Ok(latest_block) = self.get_latest_block() {
+            if block.is_parent_of(&latest_block) {
```

Comment:
This looks incorrect
```suggestion
            if !latest_block.is_parent_of(&block) {
```